cpp/src/io/json/writer_impl.cu

```diff
@@ -0,0 +1,782 @@
+ * @param append_row  Append this string to each row
+ * @param column_name_separator  Separator between column name and value
+ * @param value_separator Separator between values
+ * @param narep Null-String replacement
+ * @param include_nulls Include null string entries in the output
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource to use for device memory allocation.
+ * @return New strings column of JSON structs in each row
+ */
+std::unique_ptr<column> struct_to_strings(table_view const& strings_columns,
+                                          column_view const& column_names,
+                                          string_scalar const& prepend_row,
+                                          string_scalar const& append_row,
+                                          string_scalar const& column_name_separator,
+                                          string_scalar const& value_separator,
+                                          string_scalar const& narep,
+                                          bool include_nulls,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(column_names.type().id() == type_id::STRING, "Column names must be of type string");
+  auto const num_columns = strings_columns.num_columns();
+  // CUDF_EXPECTS(num_columns > 1, "At least two columns must be specified");
+  CUDF_EXPECTS(num_columns == column_names.size(),
+               "Number of column names should be equal to number of columns in the table");
+  // check all columns are of type string
+  CUDF_EXPECTS(std::all_of(strings_columns.begin(),
+                           strings_columns.end(),
+                           [](auto c) { return c.type().id() == type_id::STRING; }),
+               "All columns must be of type string");
+  auto const strings_count = strings_columns.num_rows();
+  if (strings_count == 0)  // empty begets empty
+    return make_empty_column(type_id::STRING);
+
+  CUDF_EXPECTS(prepend_row.is_valid(stream), "Parameter prepend_row must be a valid string_scalar");
+  CUDF_EXPECTS(append_row.is_valid(stream), "Parameter append_row must be a valid string_scalar");
+  CUDF_EXPECTS(column_name_separator.is_valid(stream),
+               "Parameter column_name_separator must be a valid string_scalar");
+  CUDF_EXPECTS(value_separator.is_valid(stream),
+               "Parameter value_separator must be a valid string_scalar");
+  string_view d_prepend_row(prepend_row.data(), prepend_row.size());
+  string_view d_append_row(append_row.data(), append_row.size());
+  string_view d_col_separator(column_name_separator.data(), column_name_separator.size());
+  string_view d_val_separator(value_separator.data(), value_separator.size());
+  auto d_narep = get_scalar_device_view(const_cast<string_scalar&>(narep));
```

Comment:
These could be removed/inlined in the `concat_structs_base` initialization.

Reply:
udpated to `string_view`